lib/llvmopencl/WorkitemLoops.cc

```diff
@@ -965,7 +965,20 @@ WorkitemLoops::GetContextArray(llvm::Instruction *instruction)
             allocType =
               ArrayType::get(elementType->getArrayElementType(), alignedSize);
           } else if (isa<StructType>(elementType)) {
-          // TODO handle structs properly
+            StructType *old_struct = dyn_cast<StructType>(elementType);
```

Comment:
Can you remind me why is this code needed and what it does exactly? Doesn't ArrayType::get() alignment argument work for structs for some reason? I think we discussed this but I forgot what was the exact problem. It should be added to comments here. You create a new type with a padding array and the old struct's elements, but what if the struct has child structs recursively etc.?

Reply:
> Can you remind me why is this code needed and what it does exactly? 

It re-aligns array (and now struct) types in WILoop's Context array. It's explained in a comment:

```
      /* 3D context array. In case the elementType itself is an array or struct,
       * we must take into account it could be alloca-ed with alignment and loads
       * or stores might use vectorized instructions expecting proper alignment.
       * Because of that, we cannot simply allocate x*y*z*(size), we must
       * enlarge the array type to fit the alignment. */
```
We've been fixing this in issue #516 already, but i only fixed arrays.

Perhaps the proper fix would be to tell LLVM to emit unaligned vector instructions, but i have no idea how to do that.

> ArrayType::get() alignment argument work for structs

There's no alignment argument in ArrayType::get(); ``alignedSize`` is the array size argument. This line of code creates a new ArrayType (with more elements).

> but what if the struct has child structs recursively etc.?

Then it may or may not break, depending on how LLVM calculates the sizes.

Right now even simple structs break on this, so i think it's an improvement. I agree padding is not the best way to fix this, but i haven't found a simple way to tell LLVM to create structs with alignments.

Reply:
OK, I see. IIRC in LLVM the alignment info is attached to the loads and stores accessing the data, not to the variable, so there might not be another way.